Add semihosting sys_exit support to riscv

t-moe opened the pull request:
I've implemented SYS_EXIT semihosting support for riscv, similar on how its done for the cortex-m.

Works with a simple test program on my esp32c6:

```
riscv32-esp-elf-objdump  -d target/riscv32imac-unknown-none-elf/debug/riscv_semihosting_example --section=.text | grep ebreak -C1
42001150:	01f01013          	slli	zero,zero,0x1f
42001154:	00100073          	ebreak
42001158:	40705013          	srai	zero,zero,0x7
```

```
probe-rs run --chip esp32c6 --format idf target/riscv32imac-unknown-none-elf/debug/riscv_semihosting_example
```




Please advise on:
* the todo regarding renabling breakpoints:  https://github.com/probe-rs/probe-rs/compare/master...t-moe:probe-rs:features/riscv_semihosting?expand=1#diff-30133d43f1cf396fea0c96eaf00b38660fa961edf89b716da104dc0288c4ae36R327

Yatekii commented on probe-rs/src/architecture/riscv/mod.rs:
```diff
@@ -115,6 +117,37 @@ impl<'probe> Riscv32<'probe> {
+        let mut reason = old_reason;
+        // Check for semihosting instruction
+        let pc: u32 = core.read_core_reg(core.program_counter().id)?.try_into()?;
+        let instruction_last = core.read_word_32((pc - 4) as u64)?;
```
It would be better to read an array of 12 bytes.

Yatekii reviewed (commented):
Thanks a lot for implementing this!

I added some comments :)

t-moe commented:
@Yatekii 
> Thanks a lot for implementing this!
> 
> I added some comments :)

Thanks for the review.
I have incorporated your requested changes as far as possible.

Please also advise on what to do with the re-enable of the breakpoints (is this in the right place where it is?)

(I'll rebase and squash my commits, once we're done here.)

Yatekii commented:
Thanks a lot! Sorry, completely forgot about the question ... i think we are good leaving it as is atm. We can always go back and revisit :)

Yatekii commented on probe-rs/src/architecture/arm/core/cortex_m.rs:
```diff
@@ -176,43 +176,37 @@ pub(crate) fn write_core_reg(
-            _ => {
-                tracing::warn!("Unknown semihosting operation r0={r0:08x} r1={r1:08x}");
+
+        let instruction2 = core.read_word_8(pc as u64)?;
```
This is still two single reads instead of a single one :)

t-moe replied:
Ok ... there you go... 
But be aware: I was not able to test it on cortex-m !!

Yatekii replied:
That's fine, thanks so much!
master can quickly be fixed and is tested a lot. It has not 100% stability guarantee and we can usually fix small issues within minutes of a problem so that's fine :)
